pkg/binarywrapper/binarywrapper_test.go

```diff
@@ -0,0 +1,277 @@
+// binarywrapper_test was chosen to be blackbox testing of the wrapper
+// to try and better understand the behavior that end consumers of the wrapper
+// might experience as they're using this. That, and there isn't much "internal"
+// to whitebox test anyhow.
+package binarywrapper_test
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"os"
+	"strings"
+	"testing"
+
+	"github.com/go-vela/vela-openssh/pkg/binarywrapper"
+	"github.com/sirupsen/logrus"
+)
+
+const (
+	testMainEnvVar        = "env-var"
+	testMainSuccessOutput = "success-output"
+	testMainFailOutput    = "fail-output"
+)
+
+// TestMain is used so that we can mock calls to binaries that need
+// to return specific output, or errors, exit codes, etc.
+func TestMain(m *testing.M) {
+	switch os.Getenv("GO_MAIN_TEST_CASE") {
+	case "":
+		os.Exit(m.Run())
+	case testMainEnvVar:
+		if strings.Contains(strings.Join(os.Args, ""), "$") {
+			os.Exit(1)
+		}
+		os.Exit(0)
+	case testMainSuccessOutput:
+		if len(os.Args) != 4 {
+			fmt.Printf("invalid os.Args: %s", strings.Join(os.Args, " "))
+			os.Exit(2)
+		}
+		fmt.Println(os.Args[2])
+		fmt.Fprint(os.Stderr, os.Args[3])
+		os.Exit(0)
+	case testMainFailOutput:
+		if len(os.Args) != 4 {
+			fmt.Printf("invalid os.Args: %s", strings.Join(os.Args, " "))
+			os.Exit(3)
+		}
+		fmt.Println(os.Args[2])
+		fmt.Fprint(os.Stderr, os.Args[3])
+		os.Exit(4)
+	}
+}
+
+type mockExecConfig struct {
+	validationError string
+	setupError      string
+	binaryPath      string
+	arguments       []string
+	environment     map[string]string
+}
+
+func (m *mockExecConfig) Validate() error {
+	if m.validationError != "" {
+		return errors.New(m.validationError)
+	}
+
+	return nil
+}
+
+func (m *mockExecConfig) Setup() error {
+	if m.setupError != "" {
+		return errors.New(m.setupError)
+	}
+
+	return nil
+}
+
+func (m *mockExecConfig) Binary() string {
+	return m.binaryPath
+}
+
+func (m *mockExecConfig) Arguments() []string {
+	if len(m.arguments) > 0 {
+		return m.arguments
+	}
+	return []string{}
+}
+
+func (m *mockExecConfig) Environment() map[string]string {
+	if m.environment != nil {
+		return m.environment
+	}
+	return map[string]string{}
+}
+
+func TestExecSuccess(t *testing.T) {
+	tests := map[string]struct {
+		config     *mockExecConfig
+		execStyle  binarywrapper.ExecStyle
+		wantStdOut string
+		wantStdErr string
+	}{
+		"formats arguments with env vars before executing": {
+			execStyle: binarywrapper.OSExecCommand,
+			config: &mockExecConfig{
+				binaryPath: os.Args[0],
+				arguments:  []string{"$SOME_TEST"},
+				environment: map[string]string{
+					"GO_MAIN_TEST_CASE": testMainEnvVar,
+					"SOME_TEST":         "Howdy!",
+				},
+			},
+		},
```

Comment:
i'm more familiar with having the first field in the struct be `name` and then using that for `t.Run` (eg: https://github.com/go-vela/server/blob/master/secret/vault/delete_test.go#L38-L51) instead of using a map. that said, we definitely have table tests without any naming, so this better than that. however, now we have a total of three styles: no name, name struct field, and this map approach. we should probably standardize.

Reply:
I'm not hard on any stance other than the fact that named tests are easier to debug when they fail. I've been leaning and experimenting more with `map[string]struct` style tests because of some of the points made on [this](https://dave.cheney.net/2019/05/07/prefer-table-driven-tests) article talking about table drive tests. Map iteration is randomized each passthrough so if there is any reliance between tests it should come out in repeated tests. Doesn't matter to me much either way though and can refactor and fix things here along the way.

Reply:
that's fair and hard to argue against. seems like a safe(r) approach, probably. something we should probably adopt across the board and somehow document for contributors as well?